### ISSUE TITLE
Fix automatic releases / snapshot releases via travis

### ISSUE THREAD

**Issue by lefou** [lihaoyi/mill#581]
This is a blocker!

All the last builds on travis failed in the release stage with the script: `ci/on-master.py ci/release.sh`. The reason is a timeout error when communicating with Sonatype OSSRH (Maven Central). Although, it could be some permission issue which is hidden to users without these permission, chances are high, releasing takes simply too long.

```
[1235/1237] scalanativelib.worker[0.3].pom 
ill.scalalib.PublishModule.publishAll 
1 targets failed
mill.scalalib.PublishModule.publishAll java.net.SocketTimeoutException: Read timed out
    sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
    sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
    sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
    java.lang.reflect.Constructor.newInstance(Constructor.java:423)
    sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1944)
    sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1939)
    java.security.AccessController.doPrivileged(Native Method)
    sun.net.www.protocol.http.HttpURLConnection.getChainedException(HttpURLConnection.java:1938)
    sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1508)
    sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
    java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
    sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:347)
    scalaj.http.HttpRequest.doConnection(Http.scala:367)
    scalaj.http.HttpRequest.exec(Http.scala:343)
    scalaj.http.HttpRequest.asString(Http.scala:491)
    mill.scalalib.publish.SonatypeHttpApi.withRetry(SonatypeHttpApi.scala:121)
    mill.scalalib.publish.SonatypeHttpApi.createStagingRepo(SonatypeHttpApi.scala:60)
    mill.scalalib.publish.SonatypePublisher.publishRelease(SonatypePublisher.scala:77)
    mill.scalalib.publish.SonatypePublisher.$anonfun$publishAll$11(SonatypePublisher.scala:55)
    mill.scalalib.publish.SonatypePublisher.$anonfun$publishAll$11$adapted(SonatypePublisher.scala:54)
    scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:789)
    scala.collection.immutable.Map$Map1.foreach(Map.scala:125)
    scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:788)
    mill.scalalib.publish.SonatypePublisher.publishAll(SonatypePublisher.scala:54)
    mill.scalalib.PublishModule$.$anonfun$publishAll$2(PublishModule.scala:118)
    scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
java.net.SocketTimeoutException: Read timed out
    java.net.SocketInputStream.socketRead0(Native Method)
    java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
    java.net.SocketInputStream.read(SocketInputStream.java:171)
    java.net.SocketInputStream.read(SocketInputStream.java:141)
    sun.security.ssl.InputRecord.readFully(InputRecord.java:465)
    sun.security.ssl.InputRecord.read(InputRecord.java:503)
    sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:983)
    sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:940)
    sun.security.ssl.AppInputStream.read(AppInputStream.java:105)
    java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
    java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
    java.io.BufferedInputStream.read(BufferedInputStream.java:345)
    sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:735)
    sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:678)
    sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1587)
    sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
    sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:263)
    scalaj.http.HttpRequest.doConnection(Http.scala:365)
    scalaj.http.HttpRequest.exec(Http.scala:343)
    scalaj.http.HttpRequest.asString(Http.scala:491)
    mill.scalalib.publish.SonatypeHttpApi.withRetry(SonatypeHttpApi.scala:121)
    mill.scalalib.publish.SonatypeHttpApi.createStagingRepo(SonatypeHttpApi.scala:60)
    mill.scalalib.publish.SonatypePublisher.publishRelease(SonatypePublisher.scala:77)
    mill.scalalib.publish.SonatypePublisher.$anonfun$publishAll$11(SonatypePublisher.scala:55)
    mill.scalalib.publish.SonatypePublisher.$anonfun$publishAll$11$adapted(SonatypePublisher.scala:54)
    scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:789)
    scala.collection.immutable.Map$Map1.foreach(Map.scala:125)
    scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:788)
    mill.scalalib.publish.SonatypePublisher.publishAll(SonatypePublisher.scala:54)
    mill.scalalib.PublishModule$.$anonfun$publishAll$2(PublishModule.scala:118)
    scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
Traceback (most recent call last):
  File "ci/on-master.py", line 10, in <module>
    subprocess.check_call(sys.argv[1:])
  File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['ci/release.sh']' returned non-zero exit status 1
The command "$CI_SCRIPT" exited with 1.
```

Sonatype OSSRH is known to be not very fast, and the count of mill release artifacts increased in the past due to new contrib packages. Currently mill tries to stage and release all packages in one go. A possible solution might be to split that step. An indicator, that that might be the right move is the fact, that the Sonatype/Nexus stage/release API can be and IS already used that way, e.g. by Maven (nexus-staging-maven-plugin) or sbt (sbt-sonatype).

This is related to #543.

**Comment by lihaoyi** [comment]
@lefou @alexarchambault is seeing a similar error when trying to publish Ammonite:

- https://github.com/lihaoyi/Ammonite/pull/946#issuecomment-477992302

Nothing changed on the Ammonite/Mill build - the Mill version didn't even change - so It could be that it's just that Sonoatype's servers are being sad.

**Comment by lihaoyi** [comment]
Or maybe related to the recent Travis CI global outage https://www.reddit.com/r/programming/comments/b6l559/travisci_has_been_down_for_over_24_hours/

**Comment by lefou** [comment]
I would rather not blame Travis. The issue was there "long" (or at least some days) before Travis was acting slow. But (only) you could exclude that by running the same task locally. I assume, Nexus just need longer than the configured timeout.

**Comment by lihaoyi** [comment]
Ammonite's build succeeded after a few more tries, on a weekend morning. Maybe we can try kicking off the travis publishing for Mill on the weekend?

**Comment by lefou** [comment]
I re-started the process on this saturday, and suddenly the Nexus deployment succeeded. The overall process failed though. Here's the error message:

```
 {"message":"Not Found","documentation_url":"https://developer.github.com/v3/repos/releases/#get-a-release-by-tag-name"}
1 targets failed
uploadToGithub java.util.NoSuchElementException: key not found: id
    scala.collection.MapLike.default(MapLike.scala:232)
    scala.collection.MapLike.default$(MapLike.scala:231)
    scala.collection.AbstractMap.default(Map.scala:59)
    scala.collection.MapLike.apply(MapLike.scala:141)
    scala.collection.MapLike.apply$(MapLike.scala:140)
    scala.collection.AbstractMap.apply(Map.scala:59)
    ujson.Js$Selector$StringSelector.apply(Js.scala:88)
    ujson.Js.apply(Js.scala:54)
    ujson.Js.apply$(Js.scala:54)
    ujson.Js$Obj.apply(Js.scala:94)
    ammonite.$file.ci.upload.apply(upload.sc:29)
    ammonite.$file.ci.upload.apply$(upload.sc:17)
    ammonite.$file.ci.upload$.apply(upload.sc:52)
    ammonite.$file.build.$anonfun$uploadToGithub$2(build.sc:590)
Traceback (most recent call last):
  File "ci/on-master.py", line 10, in <module>
    subprocess.check_call(sys.argv[1:])
  File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['ci/release.sh']' returned non-zero exit status 1
The command "$CI_SCRIPT" exited with 1.
```

**Comment by lefou** [comment]
Ok, I've found the reason for the failed GitHub upload. Travis-CI truncates the history to 50, when initially cloning the project. As we currently have more that 50 commits since the last tag, our logic to calculate the version number fails. Instead of something like: `0.3.6-68-42fcbe` it gets `42fcbe7b2d5e51b1e7a351bdd91ce96c87fdfbb2-0-42fcbe`.

**Comment by lefou** [comment]
Please note, that also the last 18 Maven Central releases got the wrong version number. E.g. `com.lihaoyi:mill-main-api_2.12:42fcbe7b2d5e51b1e7a351bdd91ce96c87fdfbb2-0-42fcbe` should be `com.lihaoyi:mill-main-api_2.12:0.3.6-68-42fcbe`

**Comment by lefou** [comment]
Fixed with commit 5d7c40ad91758d753a856abd2501cf0874ad6e96
